src/platform/plugins/shared/controls/public/controls/data_controls/options_list_control/selections_manager.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import deepEqual from 'react-fast-compare';
+import { StateComparators } from '@kbn/presentation-publishing';
+import { initializeStateManager } from '@kbn/presentation-publishing/state_manager';
+import { OptionsListControlState } from '../../../../common/options_list';
+import { OptionsListSelection } from '../../../../common/options_list/options_list_selections';
+
+function areSelectedOptionsEqual(
+  a: OptionsListSelection[] | undefined,
+  b: OptionsListSelection[] | undefined
+) {
+  return deepEqual(a ?? [], b ?? []);
+}
+
+export const selectionComparators: StateComparators<
+  Pick<OptionsListControlState, 'exclude' | 'existsSelected' | 'selectedOptions'>
+> = {
+  exclude: 'referenceEquality',
+  existsSelected: 'referenceEquality',
+  selectedOptions: areSelectedOptionsEqual,
+};
+
+export const defaultSelectionState = {
+  exclude: false,
+  existsSelected: false,
+  selectedOptions: [],
+};
+
+export type SelectionsState = Pick<
+  OptionsListControlState,
+  'exclude' | 'existsSelected' | 'selectedOptions'
+>;
+
+export function initializeSelectionsManager(
+  initialState: SelectionsState,
+  onSelectionChange: () => void
+) {
+  const selectionsManager = initializeStateManager<SelectionsState>(
+    {
+      ...initialState,
+      selectedOptions: initialState.selectedOptions ?? [],
+    },
+    defaultSelectionState
+  );
+
+  return {
+    ...selectionsManager,
+    api: {
+      ...selectionsManager.api,
+      hasInitialSelections: initialState.selectedOptions?.length || initialState.existsSelected,
+      setExclude: (next: boolean | undefined) => {
+        if (selectionsManager.api.exclude$.value !== next) {
+          selectionsManager.api.setExclude(next);
+          onSelectionChange();
+        }
+      },
+      setExistsSelected: (next: boolean | undefined) => {
```

Comment:
Is the only difference between this setter and the built-in one that it doesn't fire if the value is the same?

I wonder if we removed this extra layer if there would be any regressions. I think likely not, because how many times do we trigger a change when the value isn't actually different - and in those places, we may wrap it in an `if` statement rather than just calling it?

Reply:
I think the right place to fix this is the state manager level and the generated setter should only update the observable if the value changes. Create https://github.com/elastic/kibana/issues/221167 to track